### PR TITLE
Examples: Add post-processing volumetric fog example

### PR DESCRIPTION
**Pull request by sunag**
Related issue: -

### Description

I found a good upsampling to use in volumetric fog: https://johanneskopf.de/publications/jbu/ and I made a new example to test it.  It preserves edges while performing denoising; the images serves as a comparison against the Gaussian blur used in another volumetric example.

**[Live example](https://raw.githack.com/mrdoob/three.js/dev/examples/webgpu_postprocessing_fog.html)**

| JBU | Gaussian Blur | Disabled |
|---|---|---|
| image | image | image |

### RTT Performance

The RTT was performing many redundant calculations when used with Gaussian Blur. I am adopting the same frame update approach used in Gaussian Blur for the RTT, as well as fix the reference texture.

| Before | Now |
|---|---|
| image | image |

### CODE REVIEW

**Comment by Mugen87** [comment]
This looks nice!

Performance-wise, I have only 15 FPS on my system.

https://github.com/user-attachments/assets/fa32303f-df5b-43af-825d-285c8086cdd7

Disabling the denosier just improves the framerate to 16 FPS. So I guess the fog itself needs to be more performant.

**Comment by sunag** [comment]
> Performance-wise, I have only 15 FPS on my system.

Hmm... I was checking that out; I applied a step adjustment for raymarching. Which setting worked best for your device?

**Comment by sunag** [comment]
Performance will probably improve if I implement caching for noise; I'll check later.

**Comment by sunag** [comment]
The builds weren't up to date in the `Live example`; it should work better now.

**Comment by Mugen87** [comment]
> The builds weren't up to date in the Live example; it should work better now.

Much better! The performance is now as follows:

- JBU: 55 FPS
- Gaussian Blur: 30 FPS
- Disabled: 60 FPS

File: src/nodes/utils/RTTNode.js [hunk 1]
```diff
@@ -207,7 +207,28 @@ class RTTNode extends TextureNode {
...
+	 * Reuse the same texture reference.
+	 *
+	 * @param {Frame} frame - The frame.
+	 * @returns {Texture} The texture of the internal render target.
+	 */
+	updateReference( /*state*/ ) {
+
+		return this.renderTarget.texture;
+
+	}
```
Comment (Mugen87) [inline src/nodes/utils/RTTNode.js:235]: Um, isn't that override a no-op?

`TextureNode.updateReference()` returns `this.value` and that always points to `this.renderTarget.texture`.

Comment (sunag) [reply src/nodes/utils/RTTNode.js:235]: This issue relates to the Gaussian blur approach, which was duplicating the RTT calculation due to the reuse of `GaussianBlurNode.textureNode` a result of the `NodeFrame` reference cache. However, I’m going to update the GaussianBlur approach, it would be better to make the second pass a simple texture, since only the first pass is critical for obtaining the previous context.

**Comment by sunag** [comment]
During the reviews with Gemini, I was able to test Gaussian's energy conservation, and we corrected the coefficient array.

The test uses a white color; previously, there was slight darkening (loss of energy), but now it appears to be fully preserved.

<details>
<summary><b>Example</b></summary>

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title>three.js webgpu - gaussian blur energy test</title>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
		<link type="text/css" rel="stylesheet" href="example.css">
		<style>
			#results {
				position: absolute;
				bottom: 30px;
				left: 50%;
				transform: translateX(-50%);
				background: rgba(18, 18, 28, 0.88);
				backdrop-filter: blur(10px);
				border: 1px solid rgba(255, 255, 255, 0.15);
				border-radius: 12px;
				padding: 20px 30px;
				color: #eee;
				font-family: monospace;
				font-size: 14px;
				line-height: 1.8;
				box-shadow: 0 10px 30px rgba(0, 0, 0, 0.5);
				min-width: 380px;
				text-align: left;
			}
			.val-highlight {
				color: #00ffcc;
				font-weight: bold;
			}
			.val-match {
				color: #33ff77;
				font-weight: bold;
			}
		</style>
	</head>
	<body>

		<div id="results">
			<div><b>Input Pixel (Pure White):</b> <span id="input-val" class="val-highlight">vec4( 1.0, 1.0, 1.0, 1.0 )</span></div>
			<div><b>Output Pixel (After Blur):</b> <span id="output-val" class="val-highlight">calculating...</span></div>
			<div><b>Conservation Ratio:</b> <span id="ratio-val" class="val-match">calculating...</span></div>
			<div><b>Status:</b> <span id="status-val">...</span></div>
		</div>

		<script type="importmap">
			{
				"imports": {
					"three": "../build/three.webgpu.js",
					"three/webgpu": "../build/three.webgpu.js",
					"three/tsl": "../build/three.tsl.js",
					"three/addons/": "./jsm/"
				}
			}
		</script>

		<script type="module">

			import * as THREE from 'three/webgpu';
			import { vec4, convertToTexture, uniform } from 'three/tsl';
			import { gaussianBlur } from 'three/addons/tsl/display/GaussianBlurNode.js';
			import { GUI } from 'three/addons/libs/lil-gui.module.min.js';

			let renderer, renderPipeline;
			let blurDirectionUniform;

			const params = {
				radius: 2.0,
				sigma: 4,
				resolutionScale: 1.0
			};

			const outputValEl = document.getElementById( 'output-val' );
			const ratioValEl = document.getElementById( 'ratio-val' );
			const statusValEl = document.getElementById( 'status-val' );

			init();

			async function init() {

				renderer = new THREE.WebGPURenderer( { antialias: false } );
				renderer.setSize( window.innerWidth, window.innerHeight );
				document.body.appendChild( renderer.domElement );

				await renderer.init();

				blurDirectionUniform = uniform( params.radius );

				renderPipeline = new THREE.RenderPipeline( renderer );

				buildPipeline();

				setupGUI();

				window.addEventListener( 'resize', onWindowResize );

				renderer.setAnimationLoop( render );

			}

			let blurNode;

			function buildPipeline() {

				// 1. Create a 100% white texture node (1.0, 1.0, 1.0, 1.0)
				const whiteNode = vec4( 1.0, 1.0, 1.0, 1.0 );
				const whiteTexture = convertToTexture( whiteNode, 64, 64 );

				// 2. Apply Gaussian Blur
				blurNode = gaussianBlur( whiteTexture, blurDirectionUniform, params.sigma, {
					resolutionScale: params.resolutionScale
				} );

				// 3. Output to screen
				renderPipeline.outputNode = blurNode.renderOutput();

			}

			function setupGUI() {

				const gui = new GUI( { title: 'Energy Test Controls' } );

				gui.add( params, 'radius', 0.1, 10, 0.1 ).name( 'Blur Radius' ).onChange( ( v ) => {

					blurDirectionUniform.value = v;

				} );

				gui.add( params, 'sigma', 1, 16, 1 ).name( 'Kernel Sigma' ).onChange( () => {

					buildPipeline();

				} );

				gui.add( params, 'resolutionScale', [ 0.25, 0.5, 1.0 ] ).name( 'Res Scale' ).onChange( () => {

					buildPipeline();

				} );

			}

			let frameCount = 0;

			async function render() {

				renderPipeline.render();

				frameCount ++;

				// Read back center pixel every 10 frames
				if ( frameCount % 10 === 0 && blurNode && blurNode._verticalRT ) {

					const width = blurNode._verticalRT.width;
					const height = blurNode._verticalRT.height;
					const centerX = Math.floor( width / 2 );
					const centerY = Math.floor( height / 2 );

					const pixelData = await renderer.readRenderTargetPixelsAsync( blurNode._verticalRT, centerX, centerY, 1, 1 );

					if ( pixelData && pixelData.length > 0 ) {

						let r, g, b;

						if ( pixelData instanceof Uint16Array ) {

							// HalfFloat decoding
							r = THREE.DataUtils.fromHalfFloat( pixelData[ 0 ] );
							g = THREE.DataUtils.fromHalfFloat( pixelData[ 1 ] );
							b = THREE.DataUtils.fromHalfFloat( pixelData[ 2 ] );

						} else if ( pixelData instanceof Uint8Array ) {

							// 8-bit unorm decoding
							r = pixelData[ 0 ] / 255;
							g = pixelData[ 1 ] / 255;
							b = pixelData[ 2 ] / 255;

						} else {

							// Float32
							r = pixelData[ 0 ];
							g = pixelData[ 1 ];
							b = pixelData[ 2 ];

						}

						outputValEl.textContent = `RGBA( ${r.toFixed( 6 )}, ${g.toFixed( 6 )}, ${b.toFixed( 6 )}, 1.0000 )`;

						const percent = ( r * 100 ).toFixed( 4 );
						ratioValEl.textContent = `${percent}% (Expected: 100.0000%)`;

						if ( Math.abs( r - 1.0 ) < 0.0001 ) {

							statusValEl.textContent = 'PASSED: Exact 100.0000% Energy Conservation';
							statusValEl.style.color = '#33ff77';

						} else {

							const loss = ( ( 1.0 - r ) * 100 ).toFixed( 4 );
							statusValEl.textContent = `ENERGY LOSS: ${loss}% Darkening (Non-normalized weights)`;
							statusValEl.style.color = '#ffaa33';

						}

					}

				}

			}

			function onWindowResize() {

				renderer.setSize( window.innerWidth, window.innerHeight );

			}

		</script>
	</body>
</html>
```

</details>

Now

image

Previous

image
